arch/arc/core/dsp/swap_dsp_macros.h

```diff
@@ -0,0 +1,270 @@
+/*
+ * Copyright (c) 2022 Synopsys.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief save and load macro for ARCv2 DSP and AGU regs
+ *
+ */
+.macro _save_dsp_regs
+#ifdef CONFIG_ARC_DSP_SHARING
+	ld_s r13, [r2, ___thread_base_t_user_options_OFFSET]
+	bbit0 r13, K_DSP_IDX, dsp_skip_save
+	lr r13, [_ARC_V2_DSP_CTRL]
+	st_s r13, [sp, ___callee_saved_stack_t_dsp_ctrl_OFFSET]
+	lr r13, [_ARC_V2_ACC0_LO]
```

Comment:
From my understanding It's not entirely correct.

The ACCL/ACCH regs are used not only by 'pure' DSP instructions - they are also touched by multiply (for example) instruction which can be generated in other kernel / user code. So it's not correct to save / restore ACCL/ACCH only if `K_DSP` bit is set for thread.

Probably the best solution is to drop  ACCL/ACCH management from DSP code and select CONFIG_ARC_HAS_ACCL_REGS if CONFIG_ARC_DSP_SHARING is set.

Note that we still need to save / restore ACC0_GLO & ACC0_GHI here.

Reply:
fixed